common/src/paused_program_storage.rs

```diff
@@ -18,69 +18,61 @@
 
 use super::{program_storage::MemoryMap, *};
 use crate::storage::{MapStorage, ValueStorage};
-use gear_core::{
-    code::MAX_WASM_PAGE_COUNT,
-    pages::{GEAR_PAGE_SIZE, WASM_PAGE_SIZE},
-};
-use sp_core::MAX_POSSIBLE_ALLOCATION;
+use core::{mem, num::NonZeroU16};
+use gear_core::program::MemoryInfix;
 use sp_io::hashing;
 
-const SPLIT_COUNT: u16 = (WASM_PAGE_SIZE / GEAR_PAGE_SIZE) as u16 * MAX_WASM_PAGE_COUNT / 2;
-
 pub type SessionId = u32;
 
-// The entity helps to calculate hash of program's data and memory pages.
-// Its structure designed that way to avoid memory allocation of more than MAX_POSSIBLE_ALLOCATION bytes.
-struct Item {
-    data: (BTreeSet<WasmPage>, H256, MemoryMap),
-    remaining_pages: MemoryMap,
+pub fn hash(allocations: &BTreeSet<WasmPage>, code_hash: H256, hashes: &[H256]) -> H256 {
+    (allocations, code_hash, hashes)
+        .using_encoded(hashing::blake2_256)
+        .into()
 }
 
-impl From<(BTreeSet<WasmPage>, H256, MemoryMap)> for Item {
-    fn from(
-        (allocations, code_hash, mut memory_pages): (BTreeSet<WasmPage>, H256, MemoryMap),
-    ) -> Self {
-        let remaining_pages = memory_pages.split_off(&GearPage::from(SPLIT_COUNT));
-        Self {
-            data: (allocations, code_hash, memory_pages),
-            remaining_pages,
-        }
-    }
-}
+pub fn batch_hash(pages: &MemoryMap) -> H256 {
+    let mut data = Vec::with_capacity(sp_core::MAX_POSSIBLE_ALLOCATION as usize);
+    pages.encode_to(&mut data);
```

Comment:
Doesn't look like very effective way to calculate hash. Better to make outside big byte buffer and collect there in sequential order <page number 4 bytes> <data 16kB>, <page number 4 bytes> <data 16kB>, ....
You can read page data from right in this buffer with corresponding offset

Reply:
Let's discuss it offline

Reply:
discussed: leave that for a new task